tools/chpldef/Message.h

```diff
@@ -0,0 +1,396 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_TOOLS_CHPLDEF_MESSAGE_H
+#define CHPL_TOOLS_CHPLDEF_MESSAGE_H
+
+#include "./misc.h"
+#include "./protocol-types.h"
+#include "./Server.h"
+
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/JSON.h"
+
+#include <cstdint>
+#include <iostream>
+#include <fstream>
+
+namespace chpldef {
+
+template <typename P, typename R>
+class Request;
+class Response;
+
+/** Forward declare requests. */
+#define CHPLDEF_MESSAGE(name__, x1__, x2__, x3__) class name__;
+#include "./message-macro-list.h"
+#undef CHPLDEF_MESSAGE
+
+/**
+  Attempts to model a LSP message. A message may be either incoming or
+  outgoing (most are incoming).
+*/
+class Message {
+public:
+
+  /** Error codes are listed in order according to the LSP spec. */
+  enum Error {
+    OK                            = 0,
+    ERR_PARSE                     = -32700,
+    ERR_INVALID_REQUEST           = -32600,
+    ERR_METHOD_NOT_FOUND          = -32601,
+    ERR_INVALID_PARAMS            = -32602,
+    ERR_INTERNAL                  = -32603,
+    ERR_SERVER_NOT_INITIALIZED    = -32002,
+    ERR_UNKNOWN_ERROR_CODE        = -32001,
+    ERR_REQUEST_FAILED            = -32803,
+    ERR_SERVER_CANCELLED          = -32802,
+    ERR_CONTENT_MODIFIED          = -32801,
+    ERR_REQUEST_CANCELLED         = -32800
+  };
+
+  enum Status {
+    PENDING,        /** This message has not been handled yet. */
+    PROGRESSING,    /** This message is being handled. */
+    COMPLETED,      /** This message has been handled. */
+    FAILED          /** Handling of this message failed. */
+  };
+
+  enum Tag {
+    UNSET                 = 0,
+    INVALID               = 1,
+    RESPONSE              = 2,
+
+    // Expand the message macros.
+    #define CHPLDEF_MESSAGE(name__, x1__, x2__, x3__) name__ ,
+    #include "./message-macro-list.h"
+    #undef CHPLDEF_MESSAGE
+    NUM_MESSAGES
+  };
+
+private:
+  Tag tag_;
+  JsonValue id_;
+  Error error_ = Message::OK; 
+  Status status_ = Message::PENDING;
+  std::string note_;
+
+protected:
+  inline void markProgressing() { status_ = Message::PROGRESSING; }
+  inline void markCompleted() { status_ = Message::COMPLETED; }
+  inline void markFailed(Error error, std::string note=std::string()) {
+    status_ = Message::FAILED;  
+    error_ = error;
+    note_ = std::move(note);
+  }
+
+  static bool isIdValid(const JsonValue& id);
+
+  Message(Message::Tag tag, JsonValue&& id, Error error,
+          std::string note)
+      : tag_(tag),
+        id_(std::move(id)),
+        error_(error),
+        note_(std::move(note)) {
+    CHPL_ASSERT(isIdValid(id_));
+  }
+
+public:
+  virtual ~Message() = default;
+
+  /** Create a request given a JSON value. */
+  static chpl::owned<Message> request(Server* ctx, JsonValue&& json);
```

Comment:
Since this says "create a request", could this perhaps return a `Request`? I know `Request` is templated, but you can make a `BaseRequest` class that has all the non-template methods and code, and have each `Request` inherit from that. Then, this method can return `BaseRequest`. 

That's actually a pretty common pattern in C++. For instance, the `SmallVector<N>` extends from `SmallVectorBase` (or something of that sort), so that functions can accept any small vector and operate on it, rather than having to be templated.

Reply:
Sure, I can do that.